src/server/job.rs

```diff
@@ -1,53 +1,211 @@
 use tako::messages::common::ProgramDefinition;
 
-use crate::server::job::JobState::Waiting;
-use crate::transfer::messages::JobInfo;
-use crate::transfer::messages::JobStatus;
-use crate::TaskId;
+use crate::transfer::messages::{JobInfo, JobType, JobDetail};
+use crate::{TakoTaskId, JobTaskId, JobId, JobTaskCount, Map};
+use serde::{Deserialize, Serialize};
 
-pub type JobId = TaskId;
 
-pub enum JobState {
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub enum JobTaskState {
     Waiting,
     Running,
     Finished,
     Failed(String),
     Canceled,
 }
 
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct JobTaskInfo {
+    pub state: JobTaskState,
+    pub task_id: JobTaskId,
+}
+
+pub enum JobState {
+    SingleTask(JobTaskState),
```

Comment:
Why is it needed to distinguish these two states? Couldn't `SingleTask` just be a `Map` of size one?

Reply:
This is kind of premature optimization. I expect that some part of users will submit lots of jobs (instead of submiting job arrays) so I wanted to avoid allocation (and I did not find any smallvec alternatives to map).